modules/standard/Maps.chpl

```diff
@@ -0,0 +1,519 @@
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    /*
+      Initializes an empty map containing keys and values of given types.
+
+      :arg keyType: The type of the keys of this map.
+      :arg valType: The type of the values of this map.
+      :arg parSafe: If `true`, this map will use parallel safe operations.
+    */
+    proc init(type keyType, type valType, param parSafe=false) {
+      this.parSafe = parSafe;
+      this.keyType = keyType;
+      this.valType = valType;
+    }
+
+    /*
+      Initializes a map containing elements that are copy initialized from
+      the elements contained in another map.
+
+      :arg other: The map to initialize from.
+      :type other: map
+
+      :arg parSafe: If `true`, this map will use parallel safe operations.
+      :type parSafe: bool
+    */
+    proc init=(const ref other: map(?ps, ?kt, ?vt), param parSafe=ps) {
+      this.parSafe = parSafe;
+      this.keyType = kt;
+      this.valType = vt;
+
+      this.complete();
+
+      for key in other {
+        keys += key;
+        vals[key] = other.vals[key];
+      }
+    }
+
+    /*
+      Clears the contents of this map.
+
+      .. warning::
+
+        Clearing the contents of this map will invalidate all existing
+        references to the elements contained in this map.
+    */
+    proc clear() {
+      _enter();
+      keys.clear();
+      _leave();
+    }
+
+    /*
+      The current number of keys contained in this map.
+    */
+    inline proc const size {
+      _enter();
+      var result = keys.size;
+      _leave();
+      return result;
+    }
+
+    /*
+      Returns `true` if this map contains zero keys.
+
+      :returns: `true` if this map is empty.
+      :rtype: `bool`
+    */
+    inline proc const isEmpty(): bool {
+      return size == 0;
+    }
+
+    /*
+      Returns `true` if the given key is a member of this map, and `false`
+      otherwise.
+
+      :arg k: The key to test for membership.
+      :type k: keyType
+
+      :returns: Whether or not the given key is a member of this map.
+      :rtype: `bool`
+    */
+    proc const contains(const k: keyType): bool {
+      _enter();
+      var result = keys.contains(k);
+      _leave();
+      return result;
+    }
+
+    /*
+      Updates this map with the contents of the other, overwritting the values
+      for already-existing keys.
+
+      :arg m: The other map
+      :type m: map(keyType, valType)
+    */
+    proc update(const ref m: map(parSafe, keyType, valType)) {
+      _enter();
+      for key in m {
+        if !keys.contains(key) then
+          keys += key;
+        vals[key] = m.vals[key];
+      }
+      _leave();
+    }
+
+    /*
+      Get the value mapped to the given key, or add the mapping if key does not
+      exist.
+
+      :arg k: The key to access
+      :type k: keyType
+
+      :returns: Reference to the value mapped to the given key.
+    */
+    proc this(k: keyType) ref {
+      _enter();
```

Comment:
What's our stance on thread safety for `this`/indexing? I know that we take the stance that iterators are not required to be threadsafe, currently. Do we want to do the same for `this`?